tox_conda/plugin.py

```diff
@@ -1,332 +1,419 @@
-    # needs to be updated before being used.
-
-    # Set path to the conda executable because it cannot be determined once
-    # an env has already been created.
-    conda_exe = find_conda()
-
-    for envconfig in config.envconfigs.values():
-        # Make sure the right environment is activated. This works because we're
-        # creating environments using the `-p/--prefix` option in `tox_testenv_create`
-        envconfig.setenv["CONDA_DEFAULT_ENV"] = envconfig.setenv["TOX_ENV_DIR"]
-
-        conda_deps = [DepConfig(str(name)) for name in envconfig.conda_deps]
-        # Append filenames of additional dependency sources. tox will automatically hash
-        # their contents to detect changes.
-        if envconfig.conda_spec is not None:
-            conda_deps.append(DepConfig(envconfig.conda_spec))
-        if envconfig.conda_env is not None:
-            conda_deps.append(DepConfig(envconfig.conda_env))
-        envconfig.deps.extend(conda_deps)
-
-        envconfig.conda_exe = conda_exe
-
-
-def find_conda():
-    # This should work if we're not already in an environment
-    conda_exe = os.environ.get("_CONDA_EXE")
-    if conda_exe:
-        return conda_exe
-
-    # This should work if we're in an active environment
-    conda_exe = os.environ.get("CONDA_EXE")
-    if conda_exe:
-        return conda_exe
-
-    path = shutil.which("conda")
-
-    if path is None:
-        _exit_on_missing_conda()
-
-    try:
-        subprocess.run([str(path), "-h"], stdout=subprocess.DEVNULL)
-    except subprocess.CalledProcessError:
-        _exit_on_missing_conda()
-
-    return path
-
-
-def _exit_on_missing_conda():
-    tox.reporter.error(MISSING_CONDA_ERROR)
-    raise SystemExit(0)
-
-
-def _run_conda_process(args, venv, action, cwd):
-    redirect = tox.reporter.verbosity() < tox.reporter.Verbosity.DEBUG
-    venv._pcall(args, venv=False, action=action, cwd=cwd, redirect=redirect)
-
-
-@hookimpl
-def tox_testenv_create(venv, action):
-    tox.venv.cleanup_for_venv(venv)
-    basepath = venv.path.dirpath()
-
-    # Check for venv.envconfig.sitepackages and venv.config.alwayscopy here
-    envdir = venv.envconfig.envdir
-    python = get_py_version(venv.envconfig, action)
-
-    if venv.envconfig.conda_env is not None:
-        env_path = Path(venv.envconfig.conda_env)
+from tox.execute.api import Execute, ExecuteInstance, ExecuteOptions, ExecuteRequest, SyncWrite
+from tox.execute.local_sub_process import LocalSubProcessExecuteInstance, LocalSubProcessExecutor
+from tox.plugin import impl
+from tox.plugin.spec import EnvConfigSet, State, ToxEnvRegister
+from tox.tox_env.api import StdinSource, ToxEnvCreateArgs
+from tox.tox_env.errors import Fail
+from tox.tox_env.installer import Installer
+from tox.tox_env.python.pip.pip_install import Pip
+from tox.tox_env.python.pip.req_file import PythonDeps
+from tox.tox_env.python.virtual_env.runner import VirtualEnvRunner
+
+__all__ = []
+
+
+class CondaEnvRunner(VirtualEnvRunner):
```

Comment:
This is not a virtual environment, this shouldn't be inherting from virtualenv runner.

Reply:
I was moving back and forth on this. There are a bunch of interface functions in VirtualEnv that CondaEnv could just use as is, so I thought inheriting could be better than copying the code. But I was also uncertain whether this is a good idea or not, so I will remove the inheritance then.

Reply:
This should inherit from Python, what stuff could conda use? I don't think it should for most.

Reply:
@gaborbernat  Is it okay to use `PythonRun` from `tox` API point of view?

Reply:
I have pushed a commit to inherit from `PythonRun`. I chose this because this class provides configs like `deps` that are needed in this context as well, plus it inherits from `RunToxEnv` that provides the entire tox runner interface. @gaborbernat Let me whether this aligns with the expected architecture.